### PR TITLE
OSD: cleanup code for ms_dispatch

### PR DESCRIPTION
**Pull request by majianpeng**

### CODE REVIEW

**Comment by majianpeng** [comment]
@tchaikov . Hope your comments.

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -5864,7 +5864,6 @@ bool OSD::ms_dispatch(Message *m)
 
   do_waiters();
   _dispatch(m);
-  do_waiters();
```
Comment (liewegas) [inline src/osd/OSD.cc]: handle_osd_map drops osd_lock, so it's possible for new waiters to be queued.

**Comment by liewegas** [comment]
I think we still need the first and last patches, since _dispatch drops osd_lock in handle_osd_map (at least)

**Comment by majianpeng** [comment]
@liewegas . I'm not find osd_lock.Unlock in handle_osd_map or the related path. Could you point the place? Thanks very much!.

**Comment by majianpeng** [comment]
@tchaikov . I'm not find osd_lock.Unlock in handle_osd_map or the related path. Could you point the place?  Thanks very much!

**Comment by tchaikov** [comment]
@majianpeng neither can i find the `osd_lock.Unlock()` @liewegas mentioned. we do not drop the osd_lock since 492e6351e5723f7d97defb01d66a3505fd33a82c. maybe i am missing something. but after reviewing the code couple times, no luck.

**Comment by majianpeng** [comment]
@tchaikov . Thanks very much.

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -5856,19 +5847,10 @@ bool OSD::ms_dispatch(Message *m)
...
   }
 
-  while (dispatch_running) {
-    dout(10) << "ms_dispatch waiting for other dispatch thread to complete" << dendl;
-    dispatch_cond.Wait(osd_lock);
-  }
-  dispatch_running = true;
-
-  do_waiters();
+   do_waiters();
```
Comment (tchaikov) [inline src/osd/OSD.cc]: bad indent.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -5856,19 +5847,10 @@ bool OSD::ms_dispatch(Message *m)
...
   _dispatch(m);
 
-  dispatch_running = false;
-  dispatch_cond.Signal();
-
-  osd_lock.Unlock();
+   osd_lock.Unlock();
```
Comment (tchaikov) [inline src/osd/OSD.cc]: wrong indent.

**Comment by liewegas** [comment]
You're right--there is no more unlock.  I was looking at an older checkout.

**Comment by majianpeng** [comment]
@tchaikov  update and add 601f25b. please review. Thanks!

File: src/osd/OSD.cc (hunk 1 above)
Comment (tchaikov) [inline src/osd/OSD.cc]: @majianpeng probably we can have an assert here to make sure that the `finished` queue is empty?

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -6070,15 +6070,19 @@ void OSD::do_waiters()
   assert(osd_lock.is_locked());
 
   dout(10) << "do_waiters -- start" << dendl;
-  finished_lock.Lock();
-  while (!finished.empty()) {
-    OpRequestRef next = finished.front();
-    finished.pop_front();
-    finished_lock.Unlock();
-    dispatch_op(next);
+  //take_waiters be called in the status of osd_lock is locked.
```
Comment (tchaikov) [inline src/osd/OSD.cc]: nit: could add a space after `//`

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -6070,15 +6070,19 @@ void OSD::do_waiters()
...
+  //And !finished.empty() is very rare situation.
+  if (!finished.empty()) {
```
Comment (tchaikov) [inline src/osd/OSD.cc]: and do early return to reduce the indent level and have less code churn.

**Comment by majianpeng** [comment]
@tchaikov . update

**Comment by liewegas** [comment]
lgtm.  @athanatos do you have any concerns here?  It doesn't seem like we'll want to reintroduce an osd_lock drop.

**Comment by tchaikov** [comment]
lgtm also

File: src/osd/OSD.cc [hunk 6]
```diff
@@ -6089,6 +6071,11 @@ void OSD::do_waiters()
   assert(osd_lock.is_locked());
 
   dout(10) << "do_waiters -- start" << dendl;
+  // take_waiters be called in the status of osd_lock is locked.
+  // And !finished.empty() is very rare situation.
+  if (finished.empty())
```
Comment (athanatos) [inline src/osd/OSD.cc]: Nothing I have found suggests that std::list::empty() is safe to call with a racing write.  I think this probably invites a rare crash.

File: src/osd/OSD.cc [hunk 7]
```diff
@@ -5842,18 +5833,9 @@ bool OSD::ms_dispatch(Message *m)
...
-  }
-  dispatch_running = true;
-
   do_waiters();
   _dispatch(m);
-  do_waiters();
-
-  dispatch_running = false;
-  dispatch_cond.Signal();
+  assert(finished.empty());
```
Comment (athanatos) [inline src/osd/OSD.cc]: Same here, I don't have any reason to think that std::list::empty is safe to call with a racing write.

**Comment by athanatos** [comment]
This seems promising to me!

**Comment by majianpeng** [comment]
@athanatos . Like the comments, OSD::take_waites be called when osd_lock is locked. So those the check of finished.empty()  is the safe.

**Comment by athanatos** [comment]
@majianpeng Ah, I had assumed that it was also called from PG threads (there's an unrelated PG::take_waiters).  In that case, can we also kill finished_lock?

**Comment by majianpeng** [comment]
@athanatos  update.

**Comment by athanatos** [comment]
Code looks good, please update your commit messages to have more descriptive first lines.

As an example, the first commit could be:

osd/OSD.cc::ms_dispatch: no need to call do_waiters again after _dispatch

The second could be:

osd/OSD: remove unused take_waiter and take_waiters_front

The third could be

osd/OSD: remove dispatch_running and supporting code

The last can be

osd/OSD: remove now unnecessary finished_lock

When possible, it's good to make sure that the first line is descriptive since that's what we can see at a glance in gitk.  As a rule of thumb, "optimize" is almost never useful as a description.

**Comment by majianpeng** [comment]
@athanatos . update . Thanks your suggestions

**Comment by tchaikov** [comment]
@athanatos ping?

**Comment by tchaikov** [comment]
lgtm.

**Comment by tchaikov** [comment]
tested at 

- http://pulpito.ceph.com/kchai-2016-06-04_19:52:10-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2016-06-04_09:44:21-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2016-06-04_05:52:02-rados-wip-kefu-testing---basic-mira/

the failures are caued by environmental problem, test suite issues [1], and teuthology issue[2]

--
[1] https://github.com/ceph/ceph-qa-suite/pull/1031 
[2] http://tracker.ceph.com/issues/16142

**Pull request merged**